### PR TITLE
TcpReceiveSendGetsCanceledByDispose: update test for change in Linux kernel.

### LINKED ISSUE #91543: TcpReceiveSendGetsCanceledByDispose tests fail on Fedora 38
The following tests fail on Fedora 38:
```
System.Net.Sockets.Tests.SendReceive_SpanSyncForceNonBlocking.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True, ipv6Server: False, dualModeClient: False, owning: True)
System.Net.Sockets.Tests.SendReceive_SpanSyncForceNonBlocking.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True, ipv6Server: False, dualModeClient: True, owning: True)
System.Net.Sockets.Tests.SendReceive_SpanSyncForceNonBlocking.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True, ipv6Server: True, dualModeClient: False, owning: True)
System.Net.Sockets.Tests.SendReceive_Sync.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True, ipv6Server: False, dualModeClient: False, owning: True)
System.Net.Sockets.Tests.SendReceive_Sync.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True, ipv6Server: False, dualModeClient: True, owning: True)
System.Net.Sockets.Tests.SendReceive_Sync.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True, ipv6Server: True, dualModeClient: False, owning: True)
System.Net.Sockets.Tests.SendReceive_SyncForceNonBlocking.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True, ipv6Server: False, dualModeClient: False, owning: True)
System.Net.Sockets.Tests.SendReceive_SyncForceNonBlocking.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True, ipv6Server: False, dualModeClient: True, owning: True)
System.Net.Sockets.Tests.SendReceive_SyncForceNonBlocking.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True, ipv6Server: True, dualModeClient: False, owning: True)
System.Net.Sockets.Tests.SendReceive_SpanSync.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True, ipv6Server: False, dualModeClient: False, owning: True)
System.Net.Sockets.Tests.SendReceive_SpanSync.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True, ipv6Server: False, dualModeClient: True, owning: True)
System.Net.Sockets.Tests.SendReceive_SpanSync.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True, ipv6Server: True, dualModeClient: False, owning: True)
```

The exceptions look like:
```
System.AggregateException: System.AggregateException : One or more errors occurred. (Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)) (Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)) (Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)) (Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)) (Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)) (Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)) (Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)) (Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null))\n---- Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)\n---- Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)\n---- Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)\n---- Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)\n---- Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)\n---- Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)\n---- Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)\n---- Assert.Equal() Failure\nExpected: ConnectionReset\nActual:   (null)
        at System.RetryHelper.ExecuteAsync(Func`1 test, Int32 maxAttempts, Func`2 backoffFunc, Predicate`1 retryWhen, String testName) in /home/tester/runtime/src/libraries/Common/tests/TestUtilities/System/RetryHelper.cs:line 102
at System.Net.Sockets.Tests.SendReceive`1.TcpReceiveSendGetsCanceledByDispose(Boolean receiveOrSend, Boolean ipv6Server, Boolean dualModeClient, Boolean owning) in /home/tester/runtime/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive/SendReceive.cs:line 1032
--- End of stack trace from previous location ---
----- Inner Stack Trace #1 (Xunit.Sdk.EqualException) -----
at System.Net.Sockets.Tests.SendReceive`1.<>c__DisplayClass21_0.<<TcpReceiveSendGetsCanceledByDispose>b__0>d.MoveNext() in /home/tester/runtime/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive/SendReceive.cs:line 1120
--- End of stack trace from previous location ---
at System.RetryHelper.ExecuteAsync(Func`1 test, Int32 maxAttempts, Func`2 backoffFunc, Predicate`1 retryWhen, String testName) in /home/tester/runtime/src/libraries/Common/tests/TestUtilities/System/RetryHelper.cs:line 94
----- Inner Stack Trace #2 (Xunit.Sdk.EqualException) -----
at System.Net.Sockets.Tests.SendReceive`1.<>c__DisplayClass21_0.<<TcpReceiveSendGetsCanceledByDispose>b__0>d.MoveNext() in /home/tester/runtime/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive/SendReceive.cs:line 1120
--- End of stack trace from previous location ---
at System.RetryHelper.ExecuteAsync(Func`1 test, Int32 maxAttempts, Func`2 backoffFunc, Predicate`1 retryWhen, String testName) in /home/tester/runtime/src/libraries/Common/tests/TestUtilities/System/RetryHelper.cs:line 94
----- Inner Stack Trace #3 (Xunit.Sdk.EqualException) -----
at System.Net.Sockets.Tests.SendReceive`1.<>c__DisplayClass21_0.<<TcpReceiveSendGetsCanceledByDispose>b__0>d.MoveNext() in /home/tester/runtime/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive/SendReceive.cs:line 1120
--- End of stack trace from previous location ---
at System.RetryHelper.ExecuteAsync(Func`1 test, Int32 maxAttempts, Func`2 backoffFunc, Predicate`1 retryWhen, String testName) in /home/tester/runtime/src/libraries/Common/tests/TestUtilities/System/RetryHelper.cs:line 94
----- Inner Stack Trace #4 (Xunit.Sdk.EqualException) -----
at System.Net.Sockets.Tests.SendReceive`1.<>c__DisplayClass21_0.<<TcpReceiveSendGetsCanceledByDispose>b__0>d.MoveNext() in /home/tester/runtime/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive/SendReceive.cs:line 1120
--- End of stack trace from previous location ---
at System.RetryHelper.ExecuteAsync(Func`1 test, Int32 maxAttempts, Func`2 backoffFunc, Predicate`1 retryWhen, String testName) in /home/tester/runtime/src/libraries/Common/tests/TestUtilities/System/RetryHelper.cs:line 94
----- Inner Stack Trace #5 (Xunit.Sdk.EqualException) -----
at System.Net.Sockets.Tests.SendReceive`1.<>c__DisplayClass21_0.<<TcpReceiveSendGetsCanceledByDispose>b__0>d.MoveNext() in /home/tester/runtime/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive/SendReceive.cs:line 1120
--- End of stack trace from previous location ---
at System.RetryHelper.ExecuteAsync(Func`1 test, Int32 maxAttempts, Func`2 backoffFunc, Predicate`1 retryWhen, String testName) in /home/tester/runtime/src/libraries/Common/tests/TestUtilities/System/RetryHelper.cs:line 94
----- Inner Stack Trace #6 (Xunit.Sdk.EqualException) -----
at System.Net.Sockets.Tests.SendReceive`1.<>c__DisplayClass21_0.<<TcpReceiveSendGetsCanceledByDispose>b__0>d.MoveNext() in /home/tester/runtime/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive/SendReceive.cs:line 1120
--- End of stack trace from previous location ---
at System.RetryHelper.ExecuteAsync(Func`1 test, Int32 maxAttempts, Func`2 backoffFunc, Predicate`1 retryWhen, String testName) in /home/tester/runtime/src/libraries/Common/tests/TestUtilities/System/RetryHelper.cs:line 94
----- Inner Stack Trace #7 (Xunit.Sdk.EqualException) -----
at System.Net.Sockets.Tests.SendReceive`1.<>c__DisplayClass21_0.<<TcpReceiveSendGetsCanceledByDispose>b__0>d.MoveNext() in /home/tester/runtime/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive/SendReceive.cs:line 1120
--- End of stack trace from previous location ---
at System.RetryHelper.ExecuteAsync(Func`1 test, Int32 maxAttempts, Func`2 backoffFunc, Predicate`1 retryWhen, String testName) in /home/tester/runtime/src/libraries/Common/tests/TestUtilities/System/RetryHelper.cs:line 94
----- Inner Stack Trace #8 (Xunit.Sdk.EqualException) -----
at System.Net.Sockets.Tests.SendReceive`1.<>c__DisplayClass21_0.<<TcpReceiveSendGetsCanceledByDispose>b__0>d.MoveNext() in /home/tester/runtime/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive/SendReceive.cs:line 1120
--- End of stack trace from previous location ---
at System.RetryHelper.ExecuteAsync(Func`1 test, Int32 maxAttempts, Func`2 backoffFunc, Predicate`1 retryWhen, String testName) in /home/tester/runtime/src/libraries/Common/tests/TestUtilities/System/RetryHelper.cs:line 94
```

These tests don't fail on Fedora 37, so the behavior is possibly triggered by a change in kernel behavior.

I will investigate further when I have some time.

cc @omajid

### PR DESCRIPTION
**Pull request by tmds**
Fixes https://github.com/dotnet/runtime/issues/91543.

@antonfirsov @directhex ptal.

### CODE REVIEW

**Comment by rzikm** [comment]
oops, nevermind, wrong PR :D

**Comment by tmds** [comment]
May be also useful here, as the ppc64le/s390x jobs this was observed on are very likely not part of the regular run.

File: src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive/SendReceive.cs [hunk 1]
```diff
@@ -1031,11 +1031,12 @@ public async Task TcpReceiveSendGetsCanceledByDispose(bool receiveOrSend, bool i
...
             }
 
-            // RHEL7 kernel has a bug preventing close(AF_UNKNOWN) to succeed with IPv6 sockets.
-            // In this case Dispose will trigger a graceful shutdown, which means that receive will succeed on socket2.
-            // This bug is fixed in kernel 3.10.0-1160.25+.
-            // TODO: Remove this, once CI machines are updated to a newer kernel.
-            bool mayShutdownGraceful = UsesSync && PlatformDetection.IsRedHatFamily7 && receiveOrSend && (ipv6Server || dualModeClient);
+            // .NET uses connect(AF_UNSPEC) to abort on-going operations on Linux.
+            // Linux 6.4+ introduced a change (4faeee0cf8a5d88d63cdbc3bab124fb0e6aed08c) which disallows
+            // this operation while operations are on-going.
```
[the remark this answers is not in the document]
Comment (tmds) [reply src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive/SendReceive.cs:1036]: I think for the time being we should keep `connect(AF_UNSPEC)` as there are many kernels still that provide this, and the behavior matches closer to what Windows does.

[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive/SendReceive.cs:1036]: Eventually, though, this means our Linux implementation will not behave the way we want it to, right? Is there something better we can / should be doing instead?

[the remark this answers is not in the document]
Comment (tmds) [reply src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive/SendReceive.cs:1036]: We're doing the best we can.

I've been investigating the issue with a kernel dev. He wrote [this patch](https://marc.info/?l=linux-netdev&m=169686209917378&w=4) which fixes the regression.
We'll see if it gets included. :crossed_fingers:

**Review by antonfirsov — APPROVED** [review]

**Comment by antonfirsov** [comment]
So far it looks like this fixes the recent spike in `TcpReceiveSendGetsCanceledByDispose` failures in CI, so this should be merged as soon as the run on the PR completes. 

@dotnet/dnceng were there any recent changes in the queues where the failures started to happen? (I suspect a kernel update.)

https://dev.azure.com/dnceng-public/public/_build/results?buildId=436660&view=ms.vss-test-web.build-test-results-tab&runId=9706880&paneView=debug&resultId=209614

**Comment by directhex** [comment]
> @dotnet/dnceng were there any recent changes in the linux queues where the failures started to happen? (I suspect a kernel update.)

Ubuntu kernel 5.15.0-81.90 included a backport of the breaking change in early august - but we were probably hit by a more recent kernel package which was uploaded to the _security_ repo rather than the _updates_ repo - 5.15.0-82.91in early September, 5.15.0-83.92 in mid-September, or 5.15.0-86.96 in early October.

**Comment by antonfirsov** [comment]
The CI failures are unrelated, `TcpReceiveSendGetsCanceledByDispose` is passing everywhere.

**Pull request merged**